src/lib.rs

```diff
@@ -276,32 +276,41 @@ impl Bump {
 
     /// Allocate a new chunk and return its initialized footer.
     ///
-    /// If given, `layouts` is a tuple of the current chunk layout and the
+    /// If given, `layouts` is a tuple of the current chunk size and the
     /// layout of the allocation request that triggered us to fall back to
     /// allocating a new chunk of memory.
-    fn new_chunk(layouts: Option<(Layout, Layout)>) -> NonNull<ChunkFooter> {
+    fn new_chunk(layouts: Option<(usize, Layout)>) -> NonNull<ChunkFooter> {
         unsafe {
             let layout: Layout =
-                layouts.map_or_else(Bump::default_chunk_layout, |(old, requested)| {
-                    let old_doubled = old.size().checked_mul(2).unwrap();
+                layouts.map_or_else(Bump::default_chunk_layout, |(old_size, requested)| {
+                    let old_doubled = old_size.checked_mul(2).unwrap();
                     debug_assert_eq!(
                         old_doubled,
                         round_up_to(old_doubled, mem::align_of::<ChunkFooter>()),
                         "The old size was already a multiple of our chunk footer alignment, so no \
                          need to round it up again."
                     );
 
-                    // Round the size up to a multiple of our footer's alignment so that
-                    // we can be sure that our footer is properly aligned.
-                    let requested_size =
-                        round_up_to(requested.size(), mem::align_of::<ChunkFooter>());
+                    // Have a reasonable "doubling behavior" but ensure that if
+                    // a very large size is requested we round up to that.
+                    let size_to_allocate = cmp::max(old_doubled, requested.size());
+
+                    // Handle size/alignment of our allocated chunk, taking into
+                    // account an overaligned allocation if one is required.
+                    // Note that we also add to the size a `ChunkFooter` because
+                    // we'll be placing one at the end, and we need to at least
+                    // satisfy `requested.size()` bytes.
+                    let size = cmp::max(
```

Comment:
Bah -- I was totally aware of this and even thought I was handling this corner case already. Goes to show how nice it is to have unsafe code encapsulated in utility crates, because if I had to deal with this all the time, my code would be crashy as hell :-p